### ISSUE TITLE
Struct can not contain a pointer field

### ISSUE THREAD

**Issue by bfdy18** [googleapis/google-cloud-go#1296]
## Client

cloud.google.com/go v0.35.1

## Describe Your Environment

OSX, Using Google Datastore simulator, go version go1.11.4 darwin/amd64

## Expected Behavior

Using pointer to struct as member, example:
```
type PLSChildStruct struct {
	IntField   int      `datastore:"intField"`
	BasicSlice []string `datastore:"basicSlice"`
}

type PLSData struct {
	ID         string          `datastore:"id"`
	FlattenPtr *PLSChildStruct `datastore:"flattenPtrField,flatten"`
}
```

`PLSData` can be saved into datastore, when load back, encounter error:
datastore: cannot load field "flattenPtrField.basicSlice" into a "PLSData": no such struct field

If  define PropertyLoadSaver, 
```
func (r *PLSChildStruct) Load(ps []datastore.Property) error {
	return datastore.LoadStruct(r, ps)
}

func (r *PLSChildStruct) Save() ([]datastore.Property, error) {
	return datastore.SaveStruct(r)
}
```
`PLSData` can be loaded back, but `PLSChildStruct.BasicSlice` only has the last element, for example,
Save into datastore
```
PLSData{
	ID:         1,
	FlattenPtr: &PLSChildStruct{IntField: 6, BasicSlice: []string{"basic1", "basic2"}},
},
```
When loaded back, it becomes:
```
PLSData{
	ID:         1,
	FlattenPtr: &PLSChildStruct{IntField: 6, BasicSlice: []string{"basic2"}},
},
```

Is this by design that we can not use pointer to struct? If not, how to do it?

Thanks

**Comment by vvakame** [comment]
@jadekler I think this issue is not `type: question`.
bacause 

> `PLSData` can be saved into datastore, when load back, encounter error

We can put the entity, but can't get it.
I want to get entity or get an error when do put operation.

**Comment by poy** [comment]
@bfdy18 

## tl;dr
Remove the `flatten` tag.

## Whats going on?

So I have been doing some digging and the problem seems to be with the `flatten` tag. If you look at what that does when storing into Datastore, it more or less "skips" `PLSChildStruct`. Therefore you are storing:

```
flattenPtrField.basicSlice
["basic1","basic2"]
```

Notice that the data does not contain `basicSlice` and it is instead in the name.

In contrast, if you remove the `flatten` tag, you will have stored:
```
flattenPtrField
{"basicSlice":["basic1","basic2"]}
```

In this case, `basicSlice` is now in the data as a key name that has the value of `["basic1","basic2"]`.

Now, lets look at another example:
```go
type X0 struct {
        S string
}

type X1 struct {
        X []X0 `datastore:"x,flatten"`
}
```

Notice that `X0` is a child of `X1` and does not have a slice. Instead the slice exists in the parent (where as your example the slice existed in the child).

Saving this to Datastore results in the following:
```
x.S
["basic1","basic2"]
```

So that turns out to store the data in the same way as your example even though the data structure is completely different.

This leads me to believe we don't support the `flatten` tag when the child struct has a slice.

**Comment by bfdy18** [comment]
@poy , I think it is more than that:

`flatten` can not be used with pointer to struct, with or without slice.
```
type PLSChildStruct struct {
	IntField   int      `datastore:"intField"`
	// BasicSlice []string `datastore:"basicSlice"`
}

type PLSData struct {
	ID         string          `datastore:"id"`
	FlattenPtr *PLSChildStruct `datastore:"flattenPtrField,flatten"`
}
```
Above example without slice, still fails.

If not using pointer, `flatten` will work well with child struct with slice,
```
type PLSChildStruct struct {
	IntField   int      `datastore:"intField"`
	BasicSlice []string `datastore:"basicSlice"`
}

type PLSData struct {
	ID         string          `datastore:"id"`
	Flatten PLSChildStruct `datastore:",flatten"` // not a pointer 
}
```
Above example works.

**Comment by poy** [comment]
@bfdy18 I'm having a bit of trouble recreating your error with a pointer:

```go
type PLSChildStruct struct {
	IntField int `datastore:"intField"`
}

func (r *PLSChildStruct) Load(ps []datastore.Property) error {
	return datastore.LoadStruct(r, ps)
}

func (r *PLSChildStruct) Save() ([]datastore.Property, error) {
	return datastore.SaveStruct(r)
}

type PLSData struct {
	ID         string          `datastore:"id"`
	FlattenPtr *PLSChildStruct `datastore:"flattenPtrField,flatten"`
}

func main() {
	ctx := context.Background()

	// Create a datastore client. In a typical application, you would
	// create
	// a single client which is reused for every datastore operation.
	dsClient, err := datastore.NewClient(ctx, os.Getenv("DATASTORE_PROJECT_ID"))
	if err != nil {
		panic(err)
	}

	k := datastore.NameKey("Entity", fmt.Sprintf("id-%d", time.Now().UnixNano()), nil)
	k, err = dsClient.Put(ctx, k, &PLSData{
		ID: "1",
		FlattenPtr: &PLSChildStruct{
			IntField: 2,
		},
	},
	)
	if err != nil {
		panic(err)
	}

	var loaded PLSData
	if err := dsClient.Get(ctx, k, &loaded); err != nil {
		panic(err)
	}

	if err := json.NewEncoder(os.Stdout).Encode(&loaded); err != nil {
		panic(err)
	}
}
```

```shell
$ go run main.go
{"ID":"1","FlattenPtr":{"IntField":2}}
```

**Comment by bfdy18** [comment]
@poy , try this code, two of them failed. Define **PropertyLoadSaver** can help one of the failed case (without slice in child struct), not the other.
```
import (
	"testing"

	"cloud.google.com/go/datastore"
	"github.com/stretchr/testify/require"
)

func AssertEqual(t *testing.T, src, dst interface{}) {
	props, err := datastore.SaveStruct(src)
	require.NoError(t, err)

	datastore.LoadStruct(dst, props)
	require.NoError(t, err)

	require.Equal(t, src, dst)
}

func TestDatastoreSerialization(t *testing.T) {
	type ChildWithSlice struct {
		IntField   int
		SliceField []string
	}

	t.Run("no pointer, child has slice, passed as expected", func(t *testing.T) {
		type WithFlattenChild struct {
			StrField string
			Child    ChildWithSlice `datastore:",flatten"`
		}

		src := WithFlattenChild{
			StrField: "anything",
			Child: ChildWithSlice{
				IntField:   5,
				SliceField: []string{"s1", "s2"},
			},
		}

		var dst WithFlattenChild
		AssertEqual(t, &src, &dst)
	})

	t.Run("using pointer, child has slice, why failed?", func(t *testing.T) {
		type WithFlattenChild struct {
			StrField string
			ChildPtr *ChildWithSlice `datastore:",flatten"`
		}

		src := WithFlattenChild{
			StrField: "anything",
			ChildPtr: &ChildWithSlice{
				IntField:   5,
				SliceField: []string{"s1", "s2"},
			},
		}

		var dst WithFlattenChild
		AssertEqual(t, &src, &dst)
	})

	t.Run("using pointer, NO slice in child, why failed?", func(t *testing.T) {
		type ChildWithoutSlice struct {
			IntField int
			// SliceField []string
		}

		type WithFlattenChild struct {
			StrField string
			ChildPtr *ChildWithoutSlice `datastore:",flatten"`
		}

		src := WithFlattenChild{
			StrField: "anything",
			ChildPtr: &ChildWithoutSlice{
				IntField: 5,
			},
		}

		var dst WithFlattenChild
		AssertEqual(t, &src, &dst)
	})
}
```

Thanks

**Comment by poy** [comment]
@bfdy18 Thanks for the sample code. I'm still investigating, however I want to add one thing in case anyone else stumbles into this. Currently, your example is ignoring the error from `datastore.LoadStruct(...)`. So if you change it to:

```go
err = datastore.LoadStruct(dst, props)
```

You will get the following error:

```
datastore: cannot load field "ChildPtr.IntField" into a "ds.WithFlattenChild": no such struct field
```

I will continue to look into this and see if I can figure out why this error is occurring.

**Comment by bfdy18** [comment]
@poy thanks.
Also l suggest  `SaveStruct` should return error too if `LoadStruct` does so.

**Comment by poy** [comment]
@bfdy18 It looks like it does:

```go
props, err := datastore.SaveStruct(src)
```

**Comment by bfdy18** [comment]
@poy 
I actually mean, if `LoadStruct` returns error on a kind of `struct`, this same `struct` should fail `SaveStruct` too. It is confusing that we can save a data into datastore but unable to load them back.

**Comment by poy** [comment]
@bfdy18 Ah, I see. I agree completely!

**Comment by poy** [comment]
This CL fixes the pointer issue:

https://code-review.googlesource.com/c/gocloud/+/38550

**Comment by poy** [comment]
The pointer fix has been merged: 2b3814dbf3ff7af232c7b874e69eab0b0d66f969

**Comment by poy** [comment]
CL to fix the docs:

https://code-review.googlesource.com/c/gocloud/+/38590
